pkg/cmd/operator/webhooks_test.go

```diff
@@ -0,0 +1,358 @@
+				}
+
+				if wo.TLSConfig.GetCertificate != nil {
+					t.Errorf("expected TLSConfig.GetCertificate to be nil, got %v", reflect.ValueOf(wo.TLSConfig.GetCertificate))
+				}
+			},
+		},
+		{
+			Name: "sets dynamicCertKeyPairContent and TLSConfig.GetCertificate",
+			WebhookOptions: func() *WebhookOptions {
+				dir := t.TempDir()
+				TLSCertFile := dir + "/tls.crt"
+				TLSKeyFile := dir + "/tls.key"
+
+				if err = createFileWithContent(TLSCertFile, certs.cert); err != nil {
+					panic(err)
+				}
+				if err = createFileWithContent(TLSKeyFile, certs.privateKey); err != nil {
+					panic(err)
+				}
+
+				return &WebhookOptions{
+					TLSCertFile:                    TLSCertFile,
+					TLSKeyFile:                     TLSKeyFile,
+					InsecureGenerateLocalhostCerts: false,
+				}
+			}(),
+			ExpectedError: nil,
+			ValidateWebhookOptions: func(t *testing.T, wo *WebhookOptions) {
+				if len(wo.TLSConfig.Certificates) != 0 {
+					t.Errorf("expected TLSConfig.Certificates to be of length 0, got %d", len(wo.TLSConfig.Certificates))
+				}
+
+				if wo.dynamicCertKeyPairContent == nil {
+					t.Errorf("expected dynamicCertKeyPairContent to not be nil")
+				}
+
+				if wo.TLSConfig.GetCertificate == nil {
+					t.Errorf("expected TLSConfig.GetCertificate to not be nil")
+				}
+			},
+		},
+	}
+
+	for i := range tt {
+		test := tt[i]
+		t.Run(test.Name, func(t *testing.T) {
+			t.Parallel()
+
+			err := test.WebhookOptions.Complete()
+			if !reflect.DeepEqual(test.ExpectedError, err) {
+				t.Fatalf("expected and received errors differ: %s", cmp.Diff(test.ExpectedError, err))
+			}
+
+			if test.ValidateWebhookOptions != nil {
+				test.ValidateWebhookOptions(t, test.WebhookOptions)
+			}
+		})
+	}
+}
+
+func TestRun(t *testing.T) {
+	dynamiccertificates.FileRefreshDuration = 1 * time.Second
+
+	now := time.Now()
+	certs, err := createCertificates(now)
+	if err != nil {
+		t.Fatal("can't create certificates")
+	}
+
+	t.Run("certificates reload dynamically", func(t *testing.T) {
+		t.Parallel()
+
+		var err error
+
+		var wg sync.WaitGroup
+		defer wg.Wait()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		dir := t.TempDir()
+		TLSCertFile, TLSKeyFile := dir+"/tls.crt", dir+"/tls.key"
+
+		if err = createFileWithContent(TLSCertFile, certs.cert); err != nil {
+			t.Fatalf("can't create a file: %v", err)
+		}
+		if err = createFileWithContent(TLSKeyFile, certs.privateKey); err != nil {
+			t.Fatalf("can't create a file: %v", err)
+		}
+
+		dynamicCertKeyPairContent, err := dynamiccertificates.NewDynamicServingContentFromFiles("serving-certs-test", TLSCertFile, TLSKeyFile)
+		if err != nil {
+			panic(err)
+		}
+
+		wo := WebhookOptions{
+			TLSCertFile:                    TLSCertFile,
+			TLSKeyFile:                     TLSKeyFile,
+			Port:                           5000,
+			InsecureGenerateLocalhostCerts: false,
+			TLSConfig: tls.Config{
```

Comment:
should this use Complete?

Reply:
I suppose we could, since it's turned into an integration test anyway. Are you sure we should get rid of testing Complete then (the other comment)? We don't really cover it otherwise?

Reply:
> We don't really cover it otherwise?
We'd call it from the other unit test for Run so it will be covered, probably to a bigger extend as it will test if the completed options start the program, not just that it's non-nil. I am not against some simple and meaningful test for Complete (although not necessary).

Reply:
(also e2e tests start the program by deploying the webhook)

Reply:
I've tried creating a more reasonable table for Complete but even with custom comparators I wasn't really able to test anything meaningful in a nice way, so I've removed them altogether.